osutil/disks/disks.go

```diff
@@ -0,0 +1,288 @@
+}
+
+func udevProperties(device string) (map[string]string, error) {
+	out, err := udevadmProperties(device)
+	if err != nil {
+		return nil, osutil.OutputErr(out, err)
+	}
+	r := bytes.NewBuffer(out)
+
+	return parseUdevProperties(r)
+}
+
+func parseUdevProperties(r io.Reader) (map[string]string, error) {
+	m := make(map[string]string)
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		strs := strings.SplitN(scanner.Text(), "=", 2)
+		if len(strs) != 2 {
+			// bad udev output?
+			continue
+		}
+		m[strs[0]] = strs[1]
+	}
+
+	return m, scanner.Err()
+}
+
+// DiskFromMountPoint finds a matching Disk for the specified mount point.
+func DiskFromMountPoint(mountpoint string, opts *Options) (Disk, error) {
+	// call the unexported version that may be mocked by tests
+	return diskFromMountPoint(mountpoint, opts)
+}
+
+type disk struct {
+	major int
+	minor int
+	// fsLabelToPartUUID is a map of filesystem label -> partition uuid for now
+	// eventually this may be expanded to be more generally useful
+	fsLabelToPartUUID map[string]string
+}
+
+// diskFromMountPointImpl returns a Disk for the underlying mount source of the
+// specified mount point. For mount points which have sources that are not
+// partitions, and thus are a part of a disk, the returned Disk refers to the
+// volume/disk of the mount point itself.
+func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
+	// first get the mount entry for the mountpoint
+	mounts, err := osutil.LoadMountInfo()
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	d := &disk{}
+	var partMountPointSource string
+	// loop over the mount entries in reverse order to prevent shadowing of a
+	// particular mount on top of another one
+	for i := len(mounts) - 1; i >= 0; i-- {
+		if mounts[i].MountDir == mountpoint {
+			d.major = mounts[i].DevMajor
+			d.minor = mounts[i].DevMinor
+			partMountPointSource = mounts[i].MountSource
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
+	}
+
+	// TODO:UC20: if the mountpoint is of a decrypted mapper device, then we
+	//            need to trace back from the decrypted mapper device through
+	//            luks to find the real encrypted partition underneath the
+	//            decrypted one and thus the disk device for that partition
+
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	props, err := udevProperties(partMountPointSource)
+	if err != nil && props == nil {
+		// only fail here if props is nil, if it's available we validate it
+		// below
+		return nil, fmt.Errorf("cannot find disk for partition %s: %v", partMountPointSource, err)
+	}
+
+	// ID_PART_ENTRY_DISK will give us the major and minor of the disk that this
+	// partition originated from
+	if majorMinor, ok := props["ID_PART_ENTRY_DISK"]; ok {
```

Comment:
(nitpick) This is quite nested, maybe it could be slightly simplified by avoiding the else and returning in the if, something like:
```
         ...
	// ID_PART_ENTRY_DISK will give us the major and minor of the disk that this
	// partition originated from
	if majorMinor, ok := props["ID_PART_ENTRY_DISK"]; ok {
		maj, min, err := parseDeviceMajorMinor(majorMinor)
		if err != nil {
			// bad udev output?
			return nil, fmt.Errorf("cannot find disk for partition %s, bad udev output: %v", partMountPointSource, err)
		}
		d.major = maj
		d.minor = min
		return d, nil
	}

	// the partition is probably a volume or other non-physical disk, so
	// confirm that DEVTYPE == disk and return the maj/min for it
	if devType, ok := props["DEVTYPE"]; ok {
		if devType == "disk" {
			return d, nil
		}
		// unclear what other DEVTYPE's we should support for this
		return nil, fmt.Errorf("unsupported DEVTYPE %q for mount point source %s", devType, partMountPointSource)
	}

	return nil, fmt.Errorf("cannot find disk for partition %s, incomplete udev output", partMountPointSource)
}
``` 
diff at: https://paste.ubuntu.com/p/NKqPwPnVg7/

Reply:
Looks like a good material for a follow-up